### PR TITLE
Issue 426 refactor ghost nodes

### LINKED ISSUE #426: Refactor `add_ghost_nodes` to work with different symbols
`add_ghost_nodes` currently needs to access `y_slices` and so only works with `StateVectors` or concatenations of.
Refactor (e.g. using matrix multiplication instead) so that it can work with any kind of symbol. Needed for some of the different models.

### PR DESCRIPTION
**Pull request by tinosulzer**
# Description

- Refactor `add_ghost_nodes` to work with any symbol
- Refactor `Index`, add jacobian

Fixes #426 
Note: branches from #422 

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)

# Key checklist:

- [x] No style issues: `$ flake8`
- [x] All tests pass: `$ python run-tests.py --unit`
- [x] The documentation builds: `$ cd docs` and then `$ make clean; make html`

You can run all three at once, using `$ python run-tests.py --quick`.

## Further checks: 

- [x] Code is commented, particularly in hard-to-understand areas
- [x] Tests added that prove fix is effective or that feature works
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: tests/unit/test_spatial_methods/test_finite_volume.py [hunk 1]
```diff
@@ -959,7 +956,12 @@ def test_indefinite_integral(self):
         grad_and_div = pybamm.div(pybamm.grad(phi))
         int_grad_phi = pybamm.IndefiniteIntegral(grad_and_div, x)
         disc.set_variable_slices([phi])
-
+        disc._bcs = {
```
Comment (Scottmar93) [inline tests/unit/test_spatial_methods/test_finite_volume.py]: why do we now have boundary conditions here?

Comment (tinosulzer) [reply tests/unit/test_spatial_methods/test_finite_volume.py]: Needed so that the integrand has the right shape, will add a comment
